File: bishi/9-13VIVO/2.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <map>

using i64 = long long;  
const int P = 13331;

int main() {
    int n;
    std::cin >> n;
    std::map<i64, int> mp;           
    std::stack<std::string> st;      
    std::vector<i64> h(1, 0);      
    int cnt = 0;                  

    for (int i = 0; i < n; ++i) {
        std::string cmd;
        std::cin >> cmd;
        if (cmd == "push") {
            std::string k;
            std::cin >> k;
            st.push(k);
            for (char ch : k) {
                ++cnt;  
                i64 new_hash = h[cnt - 1] * P + ch;
                h.push_back(new_hash);
            }
            ++mp[h[cnt]];  
        } else if (cmd == "pop" && !st.empty()) {
            std::string k = st.top();
            st.pop();
            cnt -=int(k.size());
        }
    }
    
    std::cout << mp.size() << std::endl; 
    return 0;
}

File: bishi/9-13VIVO/2.py
P = 13331

def main():
    n = int(input())
    mp = {}
    st = []
    h = [0] * (n + 1)  
    cnt = 0  

    for i in range(n):
        parts = input().split()
        cmd = parts[0]
        
        if cmd == "push":
            k = parts[1]
            st.append(k)
            for j in range(len(k)):
                cnt += 1
                h[cnt] = h[cnt - 1] * P + ord(k[j])
            if h[cnt] not in mp:
                mp[h[cnt]] = 0
            mp[h[cnt]] += 1
        elif cmd == "pop" and st:
            k = st.pop()
            cnt -= len(k) 

    print(len(mp))

if __name__ == "__main__":
    main()
